File: demark/engine/tdst.py
"""TDST (TD Setup Trend) level calculation.

TDST levels serve as support/resistance derived from the bar immediately
before a Setup begins (bar 1).

- TDST Support (Buy Setup): true high of the bar before bar 1
- TDST Resistance (Sell Setup): true low of the bar before bar 1

True High = max(high, previous close)
True Low = min(low, previous close)
"""

from __future__ import annotations

import pandas as pd


def true_high(high: float, prev_close: float) -> float:
    """Calculate true high: max(high, previous close)."""
    return max(high, prev_close)


def true_low(low: float, prev_close: float) -> float:
    """Calculate true low: min(low, previous close)."""
    return min(low, prev_close)


def calc_tdst_support(bars: pd.DataFrame, setup_bar1_idx: int) -> float | None:
    """Calculate TDST Support level for a Buy Setup.

    TDST Support = true high of the bar immediately before bar 1 of the
    Buy Setup.

    Args:
        bars: DataFrame with columns [open, high, low, close] indexed by date.
        setup_bar1_idx: Integer position of bar 1 in the DataFrame.

    Returns:
        The TDST support level, or None if there is insufficient history.
    """
    if setup_bar1_idx < 1:
        return None

    pre_bar = bars.iloc[setup_bar1_idx - 1]
    if setup_bar1_idx < 2:
        # No bar before the pre-bar to get previous close; use pre-bar's high
        return float(pre_bar["high"])

    prev_close = float(bars.iloc[setup_bar1_idx - 2]["close"])
    return true_high(float(pre_bar["high"]), prev_close)


def calc_tdst_resistance(bars: pd.DataFrame, setup_bar1_idx: int) -> float | None:
    """Calculate TDST Resistance level for a Sell Setup.

    TDST Resistance = true low of the bar immediately before bar 1 of the
    Sell Setup.

    Args:
        bars: DataFrame with columns [open, high, low, close] indexed by date.
        setup_bar1_idx: Integer position of bar 1 in the DataFrame.

    Returns:
        The TDST resistance level, or None if there is insufficient history.
    """
    if setup_bar1_idx < 1:
        return None

    pre_bar = bars.iloc[setup_bar1_idx - 1]
    if setup_bar1_idx < 2:
        return float(pre_bar["low"])

    prev_close = float(bars.iloc[setup_bar1_idx - 2]["close"])
    return true_low(float(pre_bar["low"]), prev_close)

File: src/engine/tdst.ts
/**
 * TDST (TD Setup Trend) level math.
 *
 * The two reference specs disagree on the anchor:
 *   - `extreme_of_setup` — modern public DeMARK / TradingView convention.
 *     Buy Setup → resistance = highest true high across all 9 Setup bars.
 *     Sell Setup → support = lowest true low across all 9 Setup bars.
 *     **Default for v1.**
 *   - `bar_1` — canonical Jason Perl text. Anchor = the count-1 bar's
 *     true high (Buy resistance) / true low (Sell support).
 *   - `bar_before_1` — what the legacy Python codebase did. This is wrong
 *     vs. both Perl and modern DeMARK; available only for parity testing.
 *
 * "True high" / "true low" use the previous bar's close when it is more
 * extreme than the current bar's high/low, accounting for gaps. See the
 * canonical specs and `tests/engine/fixtures/canonical/tdst-*.json`.
 */

import type { Bar, Direction } from "./types";
import type { TdstAnchor } from "./config";

export function trueHigh(bars: Bar[], i: number): number {
  if (i === 0) return bars[i]!.high;
  return Math.max(bars[i]!.high, bars[i - 1]!.close);
}

export function trueLow(bars: Bar[], i: number): number {
  if (i === 0) return bars[i]!.low;
  return Math.min(bars[i]!.low, bars[i - 1]!.close);
}

export function trueRange(bars: Bar[], i: number): number {
  return trueHigh(bars, i) - trueLow(bars, i);
}

/**
 * Compute the TDST level for a completed Setup of `direction` whose
 * count-1 bar is at index `bar1Idx` and count-9 bar at `bar9Idx`.
 *
 * For a Buy Setup we return the resistance level (above price); for Sell
 * we return the support level (below price). The caller uses the level to
 * detect cancellation breaches.
 */
export function calcTdstLevel(
  bars: Bar[],
  direction: Direction,
  bar1Idx: number,
  bar9Idx: number,
  anchor: TdstAnchor,
): number {
  if (anchor === "bar_1") {
    return direction === "buy" ? trueHigh(bars, bar1Idx) : trueLow(bars, bar1Idx);
  }
  if (anchor === "bar_before_1") {
    const idx = Math.max(0, bar1Idx - 1);
    return direction === "buy" ? trueHigh(bars, idx) : trueLow(bars, idx);
  }
  // extreme_of_setup
  let extreme: number | null = null;
  for (let i = bar1Idx; i <= bar9Idx; i++) {
    const v = direction === "buy" ? trueHigh(bars, i) : trueLow(bars, i);
    if (extreme === null) {
      extreme = v;
    } else if (direction === "buy" ? v > extreme : v < extreme) {
      extreme = v;
    }
  }
  return extreme!;
}

/**
 * TDST cancellation test. Returns `true` when the Countdown should cancel.
 *   - "true_range" (modern default): Buy cancels if `trueLow > resistance`,
 *      i.e. the entire bar lifts above the resistance.
 *   - "close": Buy cancels if `close > resistance`. Legacy variant.
 */
export function tdstBreached(
  bars: Bar[],
  i: number,
  direction: Direction,
  level: number,
  test: "true_range" | "close",
): boolean {
  if (test === "close") {
    const c = bars[i]!.close;
    return direction === "buy" ? c > level : c < level;
  }
  // true_range
  return direction === "buy" ? trueLow(bars, i) > level : trueHigh(bars, i) < level;
}
